File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/rollcall/OpenRollCall.java
package com.github.dedis.popstellar.model.network.method.message.data.rollcall;

import com.github.dedis.popstellar.model.Immutable;
import com.github.dedis.popstellar.model.network.method.message.data.*;
import com.github.dedis.popstellar.model.objects.RollCall;
import com.github.dedis.popstellar.model.objects.event.EventState;
import com.google.gson.annotations.SerializedName;

/** Data sent to open a roll call */
@Immutable
public class OpenRollCall extends Data {

  @SerializedName("update_id")
  private final String updateId;

  private final String opens;

  @SerializedName("opened_at")
  private final long openedAt;

  private final String action;

  /**
   * Constructor of a data Open Roll-Call
   *
   * @param laoId id of lao
   * @param opens The 'update_id' of the latest roll call close, or in its absence, the 'id' field
   *     of the roll call creation
   * @param openedAt timestamp corresponding to roll call open
   * @param state the state in which the roll call is when this instance is created
   */
  public OpenRollCall(String laoId, String opens, long openedAt, EventState state) {
    updateId = RollCall.generateOpenRollCallId(laoId, opens, openedAt);
    this.opens = opens;
    this.openedAt = openedAt;
    if (state == EventState.CLOSED) {
      action = Action.REOPEN.getAction();
    } else {
      action = Action.OPEN.getAction();
    }
  }

  public OpenRollCall(String updateId, String opens, long openedAt, String action) {
    this.updateId = updateId;
    this.opens = opens;
    this.openedAt = openedAt;
    this.action = action;
  }

  @Override
  public String getObject() {
    return Objects.ROLL_CALL.getObject();
  }

  @Override
  public String getAction() {
    return action;
  }

  public String getUpdateId() {
    return updateId;
  }

  public String getOpens() {
    return opens;
  }

  public long getOpenedAt() {
    return openedAt;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    OpenRollCall that = (OpenRollCall) o;

    return openedAt == that.openedAt
        && java.util.Objects.equals(updateId, that.updateId)
        && java.util.Objects.equals(opens, that.opens)
        && java.util.Objects.equals(action, that.action);
  }

  @Override
  public int hashCode() {
    return java.util.Objects.hash(updateId, opens, openedAt, action);
  }

  @Override
  public String toString() {
    return "OpenRollCall{"
        + "updateId='"
        + updateId
        + '\''
        + ", opens='"
        + opens
        + '\''
        + ", openedAt="
        + openedAt
        + ", action='"
        + action
        + '\''
        + '}';
  }
}

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/rollcall/OpenRollCall.kt
package com.github.dedis.popstellar.model.network.method.message.data.rollcall

import com.github.dedis.popstellar.model.Immutable
import com.github.dedis.popstellar.model.network.method.message.data.Action
import com.github.dedis.popstellar.model.network.method.message.data.Data
import com.github.dedis.popstellar.model.network.method.message.data.Objects
import com.github.dedis.popstellar.model.objects.RollCall
import com.github.dedis.popstellar.model.objects.event.EventState
import com.github.dedis.popstellar.utility.MessageValidator
import com.google.gson.annotations.SerializedName

/** Data sent to open a roll call */
@Immutable
class OpenRollCall : Data {
  @SerializedName("update_id") val updateId: String
  val opens: String

  @SerializedName("opened_at") val openedAt: Long

  override val action: String

  /**
   * Constructor of a data Open Roll-Call
   *
   * @param laoId id of lao
   * @param opens The 'update_id' of the latest roll call close, or in its absence, the 'id' field
   *   of the roll call creation
   * @param openedAt timestamp corresponding to roll call open
   * @param state the state in which the roll call is when this instance is created
   */
  constructor(laoId: String, opens: String, openedAt: Long, state: EventState) {
    validate(laoId, "laoId", opens, openedAt)

    this.updateId = RollCall.generateOpenRollCallId(laoId, opens, openedAt)
    this.opens = opens
    this.openedAt = openedAt

    this.action =
        if (state == EventState.CLOSED) {
          Action.REOPEN.action
        } else {
          Action.OPEN.action
        }
  }

  /**
   * Constructor of a data Open Roll-Call
   *
   * @param updateId id of the update
   * @param opens The 'update_id' of the latest roll call close, or in its absence, the 'id' field
   *   of the roll call creation
   *     @param openedAt timestamp corresponding to roll call open. Must be one of
   *       ["open", "reopen"]
   */
  constructor(updateId: String, opens: String, openedAt: Long, action: String) {
    validate(updateId, "updateId", opens, openedAt)
        .elementIsOneOf(action, "action", Action.OPEN.action, Action.REOPEN.action)

    this.updateId = updateId
    this.opens = opens
    this.openedAt = openedAt
    this.action = action
  }

  override val `object`: String
    get() = Objects.ROLL_CALL.`object`

  override fun equals(other: Any?): Boolean {
    if (this === other) {
      return true
    }
    if (other == null || javaClass != other.javaClass) {
      return false
    }
    val that = other as OpenRollCall
    return openedAt == that.openedAt &&
        updateId == that.updateId &&
        opens == that.opens &&
        action == that.action
  }

  override fun hashCode(): Int {
    return java.util.Objects.hash(updateId, opens, openedAt, action)
  }

  override fun toString(): String {
    return "OpenRollCall{updateId='$updateId', opens='$opens', openedAt=$openedAt, action='$action'}"
  }

  private fun validate(
      id: String,
      idLabel: String,
      opens: String,
      openedAt: Long
  ): MessageValidator.MessageValidatorBuilder {
    return MessageValidator.verify()
        .isNotEmptyBase64(id, idLabel)
        .isNotEmptyBase64(opens, "opens")
        .greaterOrEqualThan(openedAt, 0, "openedAt")
  }
}
